docs/motivation.md

```diff
@@ -0,0 +1,550 @@
+```
+
+This gives us a flexible way to define combinations of types on the fly, with an
+intuitive syntax. For example, if we want to describe the type signature of a function
+`f` printing two numbers and then returning the result of adding them
+together, this would be
+```verona
+f(n1: Number & ToString, n2: Number & ToString) : Number
+// TODO add ToStringNumber version
+```
+
+In contrast, in a nominal system like Java, we would have to define a new
+interface `ToStringNumber` that extends both `Number` and `ToString`:
+```java
+interface ToStringNumber extends Number, ToString { ... }
+
+class C {
+    static Number f(ToStringNumber n1, ToStringNumber n2) { ... }
+}
+```
+Furthermore we would have to change the definition of every class that we want
+to implement this new trait.
+
+In Verona, a trait with multiple methods can be further simplified by breaking
+up each method into what we call an "atomic" trait and combining them with a
+conjunction. For example, the definition of `Number` above is simply syntactic sugar for
+```verona
+type Number = {
+    add(self: Self, other: Number): Number
+} & {
+    sub(self: Self, other: Number): Number
+}
+```
+
+#### Capabilities and conjunction
+In Verona, each concrete type is tagged with a *reference capability*. The reference
+capability decides which operations are permitted through and with a reference
+depending on the viewpoint. For example, a reference with capability `mut` allows us to
+read and mutate the object, while a capability `iso` is unique and opaque to all
+read/write operations.
+
+The conjunction type operator plays well together with capabilities, allowing us
+to express capability tags as part of the regular type system. For example, given a class
+`C` we can express the type of a reference to an object of class type `C` with
+reference capability `mut` as
+```verona
+type CMut = C & mut
+```
+
+By extension this means that polymorphism in Verona doesn't need special
+handling of capabilities in type parameters. A type `A[T]` parameterized on `T`
+can simply be instantiated with a type like `C & mut`. The method types of `A`
+are then able to express further restrictions on `T` that allow us to define
+operations for each instantiation.
+```verona
+type A[T] = {
+    withImm(v : T & imm) : A[T] & imm // immutable A[T] created with immutable T
+    withMut(v : T & mut) : A[T] & mut // mutable A[T] created with mutable T
+}
+```
+
+* TODO: compare with
+    - objective c: categories
+    - ruby: open classes
+    - C#: extension methods
+
+### Disjunction types
+Verona also includes disjunction types. Assuming two types `C` and `D` we can
+define the new type
+```verona
+type COrD = C | D
+```
+This describes the union of the two types.
+
+If we assume `C` and `D` above are classes, we can furthermore write pattern
+matches to distinguish between the two types dynamically.
+
+Since in Verona there is no subtyping between classes, we are able to check that
```

Comment:
Pattern matching can be completely elided here. If it's to be discussed, it needs a separate section, after everything else. And exhaustive matching is a "late concept" within pattern matching.